### PR TITLE
[sp] add box plot

### PR DESCRIPTION
**Pull request by shrey-mage**
# Summary
- see above

# Tests
Tested "primary", "secondary", and "danger" props. Sample data distribution displays correctly.

image
image

One-sided distributions also display correctly.

image

Added tooltips (cursor looks off in the GIF but it works correctly):

cc: @johnson-mage @dy46

### CODE REVIEW

**Comment by johnson-mage** [comment]
@shrey-mage you shud rebase master and try re-adding the new package to resolve the yarn.lock conflict

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  scale = 1.0,
+  width,
+  height,
+}: BoxPlotHorizontalProps) {
+  const themeContext: ThemeType = useContext(ThemeContext);
+
+  const colorMap = [
+    {
+      color: primary,
+      fill: themeContext.chart.backgroundPrimary,
```
Comment (johnson-mage) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:40]: Default to a color from the `light` theme if `themeContext` is undefined. Same for below.

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      stroke: themeContext.chart.danger,
+    },
+  ];
+
+  const colorIdx = colorMap.findIndex(el => el.color);
+  const color = colorMap[Math.max(colorIdx, 0)];
+  
+  const { min, max, outliers } = data;
+  const yMin = Math.min(...outliers) || min; 
+  const yMax = Math.max(...outliers) || max;
```
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:60]: what happens if there are only outliers in one direction?

**Review by dy46 — COMMENTED** [review]
generally looks good, do you need to add axes or labels?

**Comment by shrey-mage** [comment]
> generally looks good, do you need to add axes or labels?

I don't think for the new design we need axes/labels for box plots.

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+  height: number;
+};
+
+function BoxPlotHorizontal({
+  data,
+  primary,
+  secondary,
+  danger,
+  width,
+  height,
```
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:33]: order these

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    median: number;
+    thirdQuartile: number;
+    max: number;
+    outliers: number[];
+  },
+  primary?: boolean;
+  secondary?: boolean;
+  danger?: boolean;
+  width?: number;
+  height: number;
```
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:24]: order these

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    <svg height={height + UNIT} width={width + UNIT}>
+      <BoxPlot
+        {...data}
+        {...color}
+        boxWidth={height}
+        horizontal
+        rx={5}
+        ry={5}
+        strokeWidth={1}
+        top={UNIT/2}
```
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:77]: spacing

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx (hunk 3 above)
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:33]: order these too

**Comment by shrey-mage** [comment]
> > > generally looks good, do you need to add axes or labels?
> > 
> > 
> > I don't think for the new design we need axes/labels for box plots.
> 
> @shrey-mage Is anything supposed to be displayed on hover?

I'll add a tooltip for the 5 number summary.

**Review by dy46 — COMMENTED** [review]
make sure you order your imports and object types. also i don't think you should be adding that many changes to the `yarn.lock` file. double check that

**Comment by johnson-mage** [comment]
> @shrey-mage you shud rebase master and try re-adding the new package to resolve the yarn.lock conflict

You may need to remove the new package and re-add so the yarn.lock doesn't completely change.

**Comment by shrey-mage** [comment]
> make sure you order your imports and object types. also i don't think you should be adding that many changes to the `yarn.lock` file. double check that

@dy46 I thought this was weird too, but I just noticed the yarn.lock got wiped in one of the latest commits on master. Should I keep these changes in then?

File: mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
+import ParentSize from '@visx/responsive/lib/components/ParentSize';
+import { BoxPlot } from '@visx/stats';
+import { ThemeContext } from 'styled-components';
+import { scaleLinear } from '@visx/scale';
+import { useContext } from 'react';
```
Comment (dy46) [inline mage_ai/frontend/components/charts/BoxPlotHorizontal.tsx:5]: can you import `React` here and move it to the top of the file?

**Review by dy46 — APPROVED** [review]

**Comment by shrey-mage** [comment]
@johnson-mage @dy46 I added tooltips for five number summary and outlier values

**Comment by johnson-mage** [comment]
@shrey-mage in the gif, the tooltip seems really far from the cursor

**Review by johnson-mage — APPROVED** [review]
approved after fixing positioning of tooltip

**Pull request merged**